rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,283 @@
+
+- `kind`: Required. If present, the resulting warning will be printed as `kind: message`.
+- `message`: Required. A string message describing the issue with the package. The value should:
+  - Start with the "This package", "The application" or equivalent, or simply with the package name.
+  - Be capitalized (unless it starts with the package name).
+  - Use a period at the end.
+- `name`: Required if there are multiple values of the same `kind`. Give the issue a custom name for more easy filtering
+- `date`: Required. An ISO 8601 `yyyy-mm-dd`-formatted date from when the issue was added.
+- `urls`: Optional, list of strings. Can be used to link issues, pull requests and other related items.
+
+Other attributes are allowed. Some message kinds may specify additional required attributes.
+
+Example values:
+
+```nix
+meta.problems = [
+  {
+    name = "python2-eol";
+    kind = "deprecated";
+    message = "This package depends on Python 2, which has reached end of life.";
+    date = "1970-01-01";
+    urls = [ "https://github.com/NixOS/nixpkgs/issues/148779" ];
+  }
+  {
+    kind = "removal";
+    message = "The application has been abandoned upstream, use libfoo instead";
+    date = "1970-01-01";
+  }
+];
+```
+
+### Problem kinds
+
+At the moment, the following values for the `kind` field of a warning are known:
+
+- `removal`: The package is scheduled for removal some time in the future.
+- `deprecated`: The package has been abandoned upstream or has end of life dependencies.
+- `maintainerless`: `meta.maintainers` is empty
+- `insecure`: The package has some security vulnerabilities
+- `broken`: The package is marked as broken
+- `unsupported`: The package is not expected to build on this platform
+
+Not all values make sense for declaration in `meta.problems`: Some may be automatically generated from other `meta` attributes (for example `maintainerless`). New kinds may be added in the future. Furthermore, some kinds are expected to be present only up to once per derivation: for example, we have no use for having multiple `maintainerless` problems, and therefore also no need to give them a name in order to distinguish them.
+
+### Nixpkgs configuration
+
+The following new config option is added to nixpkgs: `config.problemHandler`. The (currently undocumented) option `config.showDerivationWarnings` will be removed.
+
+Handler values can be either `"throw"`, `"trace"` or `"ignore"`. Future values may be added in the future. The key is of the form `packageName.problemKind` or `packageName.problemName`, where `"*"` is allowed on either level as a wildcard.
+
+```nix
+problemHandler = {
+  "*" = {
+    "*" = "throw";
+    alias = "trace";
+    maintainerless = "ignore";
+  };
+  myPackage.foo = "ignore";
+};
+```
+
+If multiple rules match a given problem of a package, the most specific handler will be called:
+
+1. `pkgName.problemName`
+2. `pkgName.problemKind`
+3. `pkgName."*"`
+4. `"*".problemName`
+5. `"*".problemKind`
+6. `"*"."*"`
```

Comment:
This seems way too powerful. With just `problemHandler."*"."*" = "ignore"` users can ignore all current _and_ future package problems. Once a package gets removed (for any of the problem kinds), they then have no grace period for migration.

Are wildcards really needed? Is this mainly for the Python 2 deprecation which causes a lot of warnings or is there any other use case that needs them? And I think there's better solutions for the Python 2 problem (throw a single warning on `pkgs.python2`; make it easier to manually ignore warnings)

Reply:
Wildcards are necessary for providing sane defaults in nixpkgs, and for easily managing backwards compatibility when introducing new features. Even though I wouldn't encourage the usage of `"*"."*"` as a user, it wouldn't damn it either: Let's say I have some machines on the stable channel and don't want to deal with them more than twice a year.

Also note that more specific values will override, so one could still have `"*".insecure = "throw"` while ignoring everything else.

Reply:
> Wildcards are necessary for providing sane defaults in nixpkgs

The default in this RFC currently just maps each kind to a handler value. For that you don't need wildcards, you can just have these defaults in the code that handles the kinds.

> easily managing backwards compatibility when introducing new features

What exactly do you mean by that? Do you have an example?

> Let's say I have some machines on the stable channel and don't want to deal with them more than twice a year.

That's already the case: Things can only break when you upgrade from one stable release to the next. Backwards-incompatible changes are not allowed on stable releases.

Reply:
> For that you don't need wildcards, you can just have these defaults in the code that handles the kinds.

The goal is that people can override those defaults. I don't see how hard-coding some special values allows for that, or how it improves anything really. If your main motivation is to remove globbing because you perceive it as "too powerful", I'd disagree on that part because the most-specific matching. If you really wanted to ignore all warnings, you'd have to do `problemHandler = mkForce { "*"."*" = "ignore"; };` which is rather unlikely to happen accidentally.

> What exactly do you mean by that? Do you have an example?

Yes, for example I want to introduce a new warning kind. It's a less important one, so I'll probably want to set the default value to `trace`/`warn` or `ignore`. This also allows for transition periods where the level is low at introduction and then gradually gets increased once most of `nixpkgs` packages get fixed. All of this would not easily be possible without wildcards.

Reply:
> The goal is that people can override those defaults.

I'm wondering about the use cases though, why do people need to change the defaults?

> Yes, for example I want to introduce a new warning kind. It's a less important one, so I'll probably want to set the default value to `trace`/`warn` or `ignore`. This also allows for transition periods where the level is low at introduction and then gradually gets increased once most of `nixpkgs` packages get fixed. All of this would not easily be possible without wildcards.

There's no need to allow third-parties to specify wildcards for this though. It's totally possible to just change the default internally in nixpkgs until everything is fixed, we can do arbitrary things in CI without affecting the user interface.

Reply:
> I'm wondering about the use cases though, why do people need to change the defaults?

One use case is to opt in into stricter error messages, globally. This may be required should we migrate the source provenance tracking to use this RFC, for example. A similar use case would be to have pedantic ignore-by-default warnings/lints which are enabled only in CI.

Reply:
> One use case is to opt in into stricter error messages, globally

Ahh that sounds pretty good actually. This use case notably doesn't conflict with my worries of being able to ignore future problems. So for that to work we need something that can only _increase_ the warning level but not decrease it. This _could_ be done with wildcards and an error when you try to decrease the level, but this feels a bit hacky and too powerful to me.

How about just having something like `problemHandler.minimumLevel` that can be either `"ignore"` (the default), `"warn"` or `"error"`. The actual level of any problem will then be the maximum between `problemHandler.minimumLevel` and the hardcoded default for the specific kind.

This does allow the use case of opting into stricter error messages globally while not allowing decreasing the default strictness.

Reply:
This would not allow us to increase the default level because then suddenly the values people configured may become invalid. (Although this may be worked around with a very careful design.)

I do appreciate creating misuse resistant interfaces, however I think this one errs on the side of being too patronizing. However, we still may add such a safe guard feature later on should people desire it. (It could even make use of its own feature itself, very meta!)

Reply:
> This would not allow us to increase the default level because then suddenly the values people configured may become invalid.

I don't know what you mean by that, which default do you mean and why does it need to be changed?

> However, we still may add such a safe guard feature later on should people desire it.

That doesn't this way though. We can't be _more_ restrictive later (that's backwards incompatible, requires deprecation), we can only be _less_ restrictive later.

Reply:
Then let's go with the existing names and rules, with the additional strictness enforced in wildcards. There will be the possibility to lift that restriction in the future.

Reply:
But then again, what use cases are not covered by a `problemHandler.minimumLevel` interface that are covered by the wildcards interface?

Reply:
I don't know. Both "users may accidentally disable important warnings and we should stop them" and "users may genuinely want to do exactly that and we should allow for it" are currently equally possible but unlikely scenarios to me. I don't want to paint myself into a corner here.

Reply:
> Both "users may accidentally disable important warnings and we should stop them" and "users may genuinely want to do exactly that and we should allow for it" are currently equally possible but unlikely scenarios to me.

I'm noticing that we have distilled two separate issues here:
- Being able to decrease the level using wildcards is dangerous and should not be allowed, not now or in the future.
- Being able to increase the level using wildcards does not have any use cases that would be covered by a `problemHandler.minimumLevel`-style setting.

To me this is very clear that wildcards shouldn't be supported.

Reply:
And just to highlight it again: Problem handlers that can ignore all problems of a specific `kind` is the same as a wildcard. So to avoid the same problems, handlers should instead be based on the name of the problem.